react/features/subtitles/components/TranscriptionSubtitles.web.js

```diff
@@ -0,0 +1,93 @@
+// @flow
+
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+
+/**
+ * The type of the React {@code Component} props of
+ * {@link TranscriptionSubtitles}.
+ */
+type Props = {
+
+    /**
+     * Map of transcriptMessageID's with corresponding transcriptMessage.
+     */
+    transcriptMessages: Map<string, Object>
+};
+
+/**
+ * React {@code Component} which can display speech-to-text results from
+ * Jigasi as subtitles.
+ */
+class TranscriptionSubtitles extends Component<Props> {
+
+    /**
+     * Updates the transcription subtitles only if the Map of transcriptMessages
+     * change otherwise prevents the unnecessary re-render.
+     *
+     * @inheritdoc
+     * @param {Object} nextProps - The props passed to the component before
+     * rendering the component.
+     * @returns {boolean} - True if props of the component changes, else false.
+     */
+    shouldComponentUpdate(nextProps) {
+        return this.props.transcriptMessages !== nextProps.transcriptMessages;
+    }
+
+    /**
+     * Implements React's {@link Component#render()}.
+     *
+     * @inheritdoc
+     * @returns {ReactElement}
+     */
+    render() {
+        const paragraphs = [];
+
+        for (const [ transcriptMessageID, transcriptMessage ]
+            of this.props.transcriptMessages) {
+            let text;
+
+            if (transcriptMessage) {
+                text = `${transcriptMessage.participantName}: `;
+
+                if (transcriptMessage.final) {
+                    text += transcriptMessage.final;
+                } else {
+                    const stable = transcriptMessage.stable
+                        ? transcriptMessage.stable : '';
+                    const unstable = transcriptMessage.unstable
+                        ? transcriptMessage.unstable : '';
```

Comment:
Some sugar syntax here would be to use js fallthrough and do (transcriptMessage.unstable || ''). One observation, why is there a need to do different behavior for final, stable and unstable? Can you give me a shot example of what the text might look like? I'd like to scratch my curiosity as to why there can't exist just a "message" key for the transcriptMessage.

Reply:
To get the subtitles ASAP, we use 'interim results', which can change over time. Each result has a confidence attached to it.

Let's say I'm in the progress of saying `I hope you will have a great day`.

The process might be as follows (each line is a new interim results, confidence is added after the hyphen).
I  - 0.5
I hope we are - 0.1
I hope  - 0.9 # at this point we set "i hope to stable" 
I hope you will burn - 0.1
I hope you will have a great - 1 # now "i hope you will have a great" is marked as stable
I hope you will have a great summer - 0.4
I hope you will have a great day - MARKED AS FINAL

Reply:
Naive question: it's worthwhile to store stables vs unstable vs final instead of storing one message?

Reply:
Yes, currently we change the subtitle at every new update, gradually updating the stable version but showing the "hypothetical" unstable version. Not sure what's better for the user experience, maybe we should just update the subtitle at every new "stable" result.